src/main/java/org/spongepowered/api/item/recipe/crafting/CraftingRecipe.java

```
@@ -0,0 +1,102 @@
+import org.spongepowered.api.item.inventory.type.GridInventory;
+import org.spongepowered.api.item.recipe.Recipe;
+import org.spongepowered.api.world.World;
+
+import java.util.List;
+
+/**
+ * A CraftingRecipe represents some craftable recipe in the game.
+ *
+ * <p>It is essentially a Predicate that checks for if a recipe is valid as well
+ * as a function from a crafting matrix to a list of {@link ItemStack} (the
+ * crafting result), therefore making it an immutable interface.</p>
+ *
+ * <p>The passed in ItemGrid is usually a crafting inventory, e.g. a 2x2 or 3x3
+ * crafting matrix.</p>
+ *
+ * <p>The requirements of a CraftingRecipe can be general, they just have to
+ * eventually return a boolean given an itemgrid.</p>
+ */
+public interface CraftingRecipe extends Recipe {
+
+    /**
+     * Checks if the given {@link GridInventory} fits the required constraints
+     * to craft this {@link CraftingRecipe}.
+     *
+     * @param grid The ItemGrid to check for validity
+     * @param world The world this recipe would be used in
+     * @return True if the given input matches this recipe's requirements
+     */
+    boolean isValid(GridInventory grid, World world);
+
+    /**
+     * This method should only be called if
+     * {@link #isValid(GridInventory, World)} returns {@code true}.
+     *
+     * <p>This method is preferred over the
+     * {@link CraftingRecipe#getExemplaryResult()} method, as it customizes
+     * the result further depending on the context.</p>
+     *
+     * <p>It is advised to use the output of
+     * {@link CraftingRecipe#getExemplaryResult()}, modify it accordingly,
+     * and {@code return} it.</p>
+     *
+     * @param grid The crafting input, typically 3x3 or 2x2
+     * @return An {@link ItemStackSnapshot}
+     */
+    ItemStackSnapshot getResult(GridInventory grid);
+
+    /**
+     * This method should only be called if
+     * {@link #isValid(GridInventory, World)} returns {@code true}.
```

Again, need to define what happens if `isValid` returns `false`

---

See above.

---

its not undefined it should throw an exception

---

Well, yes, I could add a line that the method might throw a specific exception, but it does not necessarily check the validity of the input crafting grid.